components/layout_2020/lists.rs

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::context::LayoutContext;
+use crate::dom_traversal::{NodeAndStyleInfo, NodeExt, PseudoElementContentItem};
+use crate::replaced::ReplacedContent;
+use style::properties::longhands::list_style_type::computed_value::T as ListStyleType;
+use style::properties::style_structs;
+use style::values::computed::url::UrlOrNone;
+
+/// https://drafts.csswg.org/css-lists/#content-property
+pub(crate) fn make_marker<'dom, Node>(
+    context: &LayoutContext,
+    info: &NodeAndStyleInfo<Node>,
+) -> Option<Vec<PseudoElementContentItem>>
+where
+    Node: NodeExt<'dom>,
+{
+    let style = info.style.get_list();
+
+    // https://drafts.csswg.org/css-lists/#marker-image
+    let marker_image = || match &style.list_style_image {
+        UrlOrNone::Url(url) => Some(vec![
+            PseudoElementContentItem::Replaced(ReplacedContent::from_image_url(
+                info.node, context, url,
+            )?),
+            PseudoElementContentItem::Text(" ".into()),
+        ]),
+        UrlOrNone::None => None,
+    };
+    marker_image().or_else(|| {
+        Some(vec![PseudoElementContentItem::Text(
```

Comment:
why not just inline this into the above closure?

Reply:
The closure uses `?` so it has two paths that can return `None`. This implements “If the `<image>` represents a valid image” in https://drafts.csswg.org/css-lists/#marker-image